stablehlo/dialect/TypeInference.cpp

```diff
@@ -97,6 +97,54 @@ bool tensorsHaveSameElType(Type type1, Type type2,
   return tensorsHaveSameElType({type1, type2}, ignoreFpPrecision);
 }
 
+unsigned potentiallyComplexBitWidth(Type type) {
+  auto complexTy = type.dyn_cast<ComplexType>();
+  return complexTy ? 2 * complexTy.getElementType().getIntOrFloatBitWidth()
+                   : type.getIntOrFloatBitWidth();
+}
+
+template <typename T>
+bool matchesType(Type a, Type b) {
+  return a.isa<T>() && b.isa<T>();
+}
+
+// Returns true if the element-type of type1 can be promoted to that of type2.
+// An element-type 'x' is promotatble to element-type 'y' is they have the same
+// base type and bitwidth(x) <= bitwidth(y). When 'x' and 'y' are quantized
+// element-types, then promotion is applied only to the 'storage_type'
+// component.
+bool isPromotableElementType(Type type1, Type type2,
+                             bool ignoreFpPrecision = false) {
+  auto tensorTy1 = type1.dyn_cast<ShapedType>();
+  auto tensorTy2 = type2.dyn_cast<ShapedType>();
+
+  if (!tensorTy1 || !tensorTy2) return false;
+
+  Type tensorEl1 = tensorTy1.getElementType();
+  Type tensorEl2 = tensorTy2.getElementType();
+
+  if (ignoreFpPrecision && tensorEl1.isa<FloatType>() &&
+      tensorEl2.isa<FloatType>())
+    return true;
+
+  bool isSameType = matchesType<IntegerType>(tensorEl1, tensorEl2) ||
+                    matchesType<FloatType>(tensorEl1, tensorEl2) ||
+                    matchesType<ComplexType>(tensorEl1, tensorEl2) ||
+                    matchesType<quant::QuantizedType>(tensorEl1, tensorEl2);
+
+  if (!isSameType) return false;
+
+  if (!tensorEl1.isa<quant::QuantizedType>())
+    return potentiallyComplexBitWidth(tensorEl1) <=
+           potentiallyComplexBitWidth(tensorEl2);
+
+  auto quantType1 = tensorEl1.cast<quant::QuantizedType>();
+  auto quantType2 = tensorEl2.cast<quant::QuantizedType>();
+  return quantType1.getExpressedType() == quantType2.getExpressedType() &&
+         potentiallyComplexBitWidth(quantType1.getStorageType()) <=
+             potentiallyComplexBitWidth(quantType2.getStorageType());
```

Comment:
Let's refactor the above function:

```
unsigned getBitWidth(Type type) {
  if (auto auto complexTy = type.dyn_cast<ComplexType>())
    return getBitWidth(complexTy.getElementType());
  if (auto auto quantTy = type.dyn_cast<quant::QuantizedType>())
    return getBitWidth(complexTy.getStorageType());
  return type.getIntOrFloatBitWidth();
}
```

Then we can simplify this function to:

```
return getBitWidth(tensorEl1) <= getBitWidth(tensorEl2);
```